src/Stratis.Bitcoin.Features.SmartContracts.Tests/MethodParameterByteSerializerTests.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using NBitcoin;
+using Stratis.Bitcoin.Features.SmartContracts.Networks;
+using Stratis.SmartContracts;
+using Stratis.SmartContracts.Executor.Reflection.Serialization;
+using Xunit;
+
+namespace Stratis.Bitcoin.Features.SmartContracts.Tests
+{
+    public class MethodParameterByteSerializerTests
+    {
+        public static Network Network = new SmartContractPosRegTest();
+        public IMethodParameterSerializer Serializer = new MethodParameterByteSerializer(new ContractPrimitiveSerializer(Network));
+
+        [Theory]
+        [MemberData(nameof(GetData), parameters: 1)]
+        public void Roundtrip_Method_Param_Successfully(object value)
+        {
+            // Roundtrip serialization
+            var methodParamObjects = this.Serializer.Deserialize(this.Serializer.Serialize(new[] { value }));
+
+            Type paramType = value.GetType();
+
+            // Equality comparison using .Equal is possible for these Types
+            if (paramType.IsValueType || paramType == typeof(uint160) || paramType == typeof(string))
+            {
+                Assert.Equal(value, methodParamObjects[0]);
+            }
+
+            // For byte arrays we must compare each element
+            if (paramType.IsArray && paramType.GetElementType() == typeof(byte))
+            {
+                Assert.True(((byte[])value).SequenceEqual((byte[])methodParamObjects[0]));
+            }
+        }
+
+        [Fact]
+        public void Serialized_Method_Params_Are_Smaller_Than_Strings()
+        {
+            // Single comparative case for a sample byte vs. string encoded method params array
+            var stringSerializer = new MethodParameterStringSerializer();
+
+            var parameters = GetData(0).SelectMany(o => o).ToArray();
+
+            var serializedBytes = this.Serializer.Serialize(parameters);
+            var s = stringSerializer.Serialize(parameters);
+            var serializedString = Encoding.UTF8.GetBytes(s);
+            
+            Assert.True(serializedBytes.Length <= serializedString.Length);
+        }
+
+        [Fact]
+        public void Roundtrip_Serialize_Multiple_Params()
+        {
+            object[] methodParameters = GetData(0).SelectMany(o => o).ToArray();
+
+            var serialized = this.Serializer.Serialize(methodParameters);
+
+            var deserialized = this.Serializer.Deserialize(serialized);
```

Comment:
Assert length of deserialized == methodParameters or similar here, just in case it returns an empty array, test would still pass